src/components/ActionPanel/hooks.ts

```diff
@@ -0,0 +1,177 @@
+import React from 'react';
+
+type AnimationStateType = {
+    domPending: boolean;
+    cssTransitionPending: boolean;
+};
+
+type AnimationStateActions =
+    | {
+          type: 'domPending';
+          payload: boolean;
+      }
+    | {
+          type: 'cssTransitionPending';
+          payload: boolean;
+      };
+
+const animationStateReducer = (state: AnimationStateType, action: AnimationStateActions) => {
+    switch (action.type) {
+        case 'domPending': {
+            return {
+                ...state,
+                domPending: action.payload,
+            };
+        }
+
+        case 'cssTransitionPending': {
+            return {
+                ...state,
+                // if transition started than element is ready
+                domPending: false,
+                cssTransitionPending: action.payload,
+            };
+        }
+
+        default:
+            return state;
+    }
+};
+
+export const useAnimationState = ({
+    element,
+    enable,
+    triggerProperty,
+}: {
+    element: HTMLElement | null;
+    enable: boolean;
+    triggerProperty: boolean;
+}) => {
+    const cssTransitionTimeout = React.useRef<ReturnType<typeof setTimeout>>();
+    const [state, dispatch] = React.useReducer(animationStateReducer, {
+        domPending: false,
+        cssTransitionPending: false,
+    });
+
+    const transitionHandler = React.useCallback(
+        (e?: TransitionEvent | null) => {
+            const payload = e !== null && e?.type !== 'transitionend';
+            dispatch({
+                type: 'cssTransitionPending',
+                payload,
+            });
+
+            if (payload) {
+                clearTimeout(cssTransitionTimeout.current);
+            }
+        },
+        [dispatch, cssTransitionTimeout],
+    );
+
+    const setDomPending = React.useCallback(
+        (payload: boolean) => {
+            dispatch({
+                type: 'domPending',
+                payload,
+            });
+        },
+        [dispatch],
+    );
+
+    const propertyHandler = React.useCallback(() => {
+        if (!enable) {
+            return;
+        }
+
+        setDomPending(true);
+
+        // timeout timer for transitionstart if no css transition rule
+        cssTransitionTimeout.current = setTimeout(() => {
+            setDomPending(false);
+        }, 100);
+    }, [enable, cssTransitionTimeout, setDomPending]);
+    usePropWatcher(triggerProperty, propertyHandler);
+
+    React.useEffect(() => {
+        if (!enable) {
+            element?.removeEventListener('transitionend', transitionHandler);
```

Comment:
For me it looks strange because you remove listeners not inside `useEffect` return function